### PR TITLE
Boluscalc

### PR DESCRIPTION
**Pull request by DunePlodder**

### CODE REVIEW

**Comment by stavlor** [comment]
Will test this but it's best if this goes to dev, think master is just for releases/ needed hotfixes

**Comment by DunePlodder** [comment]
I wanted to try boluscalc but I don't really know what I'm doing in Github so have probably done the wrong thing - my apologies. Please ignore/remove my request if necessary.

File: static/js/ui-utils.js [hunk 1]
```diff
@@ -420,3 +426,95 @@ $(function() {
...
+Date.prototype.toDateInputValue = (function() {
+	var local = new Date(this);
+	local.setMinutes(this.getMinutes() - this.getTimezoneOffset());
+	return local.toJSON().slice(0,10);
+});
+
+Date.prototype.addMinutes = function(h) {    
+   this.setTime(this.getTime() + (h*60*1000)); 
+   return this;   
+}
```
Comment (jasoncalabrese) [inline static/js/ui-utils.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498841/issues/source?bid=2452374&fileBranchId=2459081#l450)

File: lib/currentProfile.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+			if (parseInt(record.targetBG[i+1].from)>minutes) return convertUnits(record.targetBG[i].high,units);
+		}
+		return null;
+	}
+	
+	function convertUnits(value, units) { // convert value to specified units. if no units specified return in mg/dL
+		value = parseFloat(value);
+		if (units.toLowerCase() == 'mg/dl' || typeof units === 'undefined') {
+			if (record.units == 'mmol') return value * 18;
+			else return value;
```
Comment (jasoncalabrese) [inline lib/currentProfile.js]: Codacy Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498805/issues/source?bid=2452374&fileBranchId=2459081#l101)

File: static/js/reports/report.js [hunk 3]
```diff
@@ -0,0 +1,1051 @@
...
+        charts = d3.select('#'+containerprefix+day).html(
+			'<b>'+
+			localeDate(day)+
+			'</b><br>'
+		).append('svg');
+
+		charts.append("rect")
+			.attr("width", "100%")
+			.attr("height", "100%")
+			.attr("fill", "WhiteSmoke");
```
Comment (jasoncalabrese) [inline static/js/reports/report.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498838/issues/source?bid=2452374&fileBranchId=2459081#l153)

File: static/js/food.js [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+	var icon_remove = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAACrElEQVQ4T42Ty2sTQRzHv5tmk2yyjRNtpfZhL8V6s2KoUNC2XqwgaCsVQcGiFqpHi0c9iRdR/ANE9KR40FIQX4cueKoPaKFoLdSYNtE0abKT1+5s9iW7aUMiHtzTzO7v85md+c6PA4DrHbsPCKIgOWO1pA7dT6YXnXH949SE/F63pqwZtRrO+SCKgjQ5NUV+azpmHj2krMwaJC4c8Erj+/eRyloMMwWFKgbn1nC3ervlK1evkXBLGBZT8SOewotnTylTNLdgeg/pDgZDC2cPHSR8bB22DVC9hFe0SG/H0xFXcHlykjRHRDBWgJcZSCY38Xx2lhqMnRYE34Px/sN9vlQWeoHBAx2yXsRruVAVuFsIBaSJ8+eJGPaBqQV4NROJjTzez89jLBoFn6FgybQL54wS3uTyVDFQ3cL2IYpBv3RhdJSIIQ80tQyv7gEqJvS8AmUlBs7UXPhtjtZgh3UFNYngk86NHCfNAg9dMwHVBPu+CpsVkTXKeJeVG+AGgTOZ3tt6MSKKjy+NjEBjFrR4ElZmA4pdxstMFsyyJu6tZZ7Ux9vwB6EAL50ZGiRECEPPUOixVTRxHlicgSVWxEdZpuZWfNuS2hk48NjwMIkIYZglBnV5Cbqtws/5IaAJmsfCglrEl2y2QeKmEBJ80tixKmxrFpSVr0gV0viQoxho2YUuPohmeFD22PiklLC4ma5JuBvdrfLJI0dJd0s7bM0ES8aR/BXDXGaTskqlL+D3Lwy0tZEePoAd4EA5YF4tYymdonfjmQh3s6dTPjU4SHYGwjAKecSXFyGlM1TdytntE56T+ts7SC/vhw3gm6njc2Kd3vm5Ub1IwQAvnYhGiZpYw1wiWYPrIw7wnBTt7CLOOwdmut14kQQvqt24tfK/utGR6LaF+iRqMf4N/O/8D28HiiCRYqzAAAAAAElFTkSuQmCC";
```
Comment (jasoncalabrese) [inline static/js/food.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498828/issues/source?bid=2452374&fileBranchId=2459081#l59)

File: static/js/hashauth/hashauth.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+									alert(translate('Wrong API secret'));
+								}
+								$( dialog ).dialog( "close" );
+							});
+							return false;
+						}
+					}
+						
+				},
+				{ text: translate("Remove"),
```
Comment (jasoncalabrese) [inline static/js/hashauth/hashauth.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498829/issues/source?bid=2452374&fileBranchId=2459081#l76)

File: lib/currentProfile.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		}
+		if (units == 'mmol') {
+			if (record.units == 'mmol') return value;
```
Comment (jasoncalabrese) [inline lib/currentProfile.js]: Codacy Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498805/issues/source?bid=2452374&fileBranchId=2459081#l104)

File: static/js/reports/report.js [hunk 7]
```diff
@@ -0,0 +1,1051 @@
...
+	var targetBGdefault = {
+		"mg/dl": { low: 72, high: 180 },
+		"mmol": { low: 4, high: 10 }
+	};
+	
+	var maxInsulinValue = 0
+		, maxCarbsValue = 0;
+    var 
+          ONE_MIN_IN_MS = 60000
+        , SIX_MINS_IN_MS =  360000
```
Comment (jasoncalabrese) [inline static/js/reports/report.js]: Codacy Issue found: ['SIX_MINS_IN_MS' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498838/issues/source?bid=2452374&fileBranchId=2459081#l20)

File: static/js/reports/calibrations.js [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+	for (var i=events.length-1; i>0; i--) {
+		if (typeof events[i].device !== 'undefined') {
+			events[i].checked = true;
+			$('#calibrations-'+i).prop('checked',true);
+			if (--maxcals<1) break;
+		}
+	}
+	calibrations_drawelements();
+
+	$('.calibrations-checkbox').change(calibrations_checkboxevent);
```
Comment (jasoncalabrese) [inline static/js/reports/calibrations.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498832/issues/source?bid=2452374&fileBranchId=2459081#l72)

File: static/js/reports/calibrations.js [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+		} else if (typeof e.scale !== 'undefined') {
+			html += '<b style="padding-left:4em">CAL</b>: ' + ' Scale: ' + e.scale.toFixed(2) + ' Intercept: ' + e.intercept.toFixed(0) + ' Slope: ' + e.slope.toFixed(2) + '<br>';
+			if (lastmbg) lastmbg.cals.push(e);
+		} else html += JSON.stringify(e);
+		html += '</td></tr>';
+	};
+	
+	html += '</table>';
+
+	$('#calibrations-list').html(html);
```
Comment (jasoncalabrese) [inline static/js/reports/calibrations.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498832/issues/source?bid=2452374&fileBranchId=2459081#l59)

File: lib/currentProfile.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+			if (parseInt(record.isf[i+1].from)>minutes) return convertUnits(record.isf[i].val,units);
+		}
+		return null;
+	}
+
+	function car(time, glycemicIndex) { //glycemicIndex 1..low  2..medium 3..high but allowing float within range <1-3>
+		if (!record) return null;
+		if (record.simple) return record.carbs_hr;
+		if (glycemicIndex==1) return record.car_low;
+		if (glycemicIndex==2) return record.car_medium;
```
Comment (jasoncalabrese) [inline lib/currentProfile.js]: Codacy Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/769498805/issues/source?bid=2452374&fileBranchId=2459081#l52)

**Comment by jasoncalabrese** [comment]
We'll get the bolus calc in, but as a plugin, still some more refactoring to do, but getting closer

**Comment by jasoncalabrese** [comment]
@DunePlodder if you haven't tried `dev` yet you should give it a try, it doesn't have the full bolus calculator yet, but does show the BWP (bolus wizard preview) and some other plugins.  If you need it there are some instructions for setting up a test site here: https://github.com/nightscout/cgm-remote-monitor/wiki/Deploy-a-test-branch

**Comment by DunePlodder** [comment]
Hi Jason,
I'm testing the rawbg-units branch with BWP at the moment. Will look at 
the Wiki as you suggest - I do feel a bit lost on Github.
Thank you.
Regards,
Robert

**Comment by jasoncalabrese** [comment]
That branch was merged to dev a while ago asking with several others

**Pull request closed without merging**